api/mongowriter.go

```diff
@@ -1,26 +1,70 @@
 package api
 
 import (
+	"bytes"
+	"context"
 	"net/http"
+	"strings"
 
 	"github.com/sirupsen/logrus"
-	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
-// MongoWriter is a custom http.ResponseWriter to commit or abort transactions.
-type MongoWriter struct {
-	logger *logrus.Logger
-	sctx   mongo.SessionContext
-	w      http.ResponseWriter
-}
+const (
+	// writeConflictErrMsg is the error message MongoDB issues when a
+	// transaction needs to be reverted because of a write conflict.
+	writeConflictErrMsg = "(WriteConflict) WriteConflict error: this operation conflicted with another operation. Please retry your operation or multi-document transaction."
```

Comment:
not sure whether this is the safe thing to do, if they roll out a driver update that slightly changes copy here our code breaks and it'll go unnoticed, is there an option to use the `WriteConcernError` from the driver package and analyse the `WriteErrors` individually to detect the write conflict?

Reply:
yeah that's what I meant, you can cast the error as a mongo type though and then parse through its write errors and look for `WriteConflict`, but I think just doing the `strings.Contains` is probably fine also